### PR TITLE
Introduce StatusOptions.IncludeUnaltered

### PR DESCRIPTION
**Pull request by Therzok**
Test can be improved, what else needs done?

### CODE REVIEW

**Comment by ethomson** [comment]
@Therzok There's a libgit2 bug in here around combining the two lists.  I'll try to take a look at it further this week.  Sorry about the delay with the holidays and the $DAYJOB and such.

**Comment by Therzok** [comment]
Okay, this seems fixed in latest libgit2. Merge at will. The appveyor failure seems unrelated.

File: LibGit2Sharp/RepositoryStatus.cs [hunk 1]
```diff
@@ -81,7 +82,7 @@ internal RepositoryStatus(Repository repo, StatusOptions options)
                     AddStatusEntryForDelta(entry.Status, deltaHeadToIndex, deltaIndexToWorkDir);
                 }
 
-                isDirty = statusEntries.Any(entry => entry.State != FileStatus.Ignored);
+                isDirty = statusEntries.Any(entry => entry.State != FileStatus.Ignored && entry.State != FileStatus.Unaltered);
```
Comment (nulltoken) [inline LibGit2Sharp/RepositoryStatus.cs]: @Therzok As we're modifying the behavior of the `IsDirty` property, could you please add some test coverage regarding this?

Comment (Therzok) [inline LibGit2Sharp/RepositoryStatus.cs]: Actually, we don't quite modify the behaviour. We extend it here.

Without IncludeUnaltered, we never had Unaltered file status. I'll add coverage for it though.

Comment (Therzok) [inline LibGit2Sharp/RepositoryStatus.cs]: That said, does having Untracked files mark the status as dirty? Thoughts?

Comment (nulltoken) [inline LibGit2Sharp/RepositoryStatus.cs]: I'd say no. What does git.git has to say about this?

Comment (Therzok) [inline LibGit2Sharp/RepositoryStatus.cs]: I just did a:

`test -n "$(git status --porcelain)" || echo "asdf"` with untracked files and that makes it considered dirty. "asdf" isn't printed.

Okay, should be good to go now.

File: LibGit2Sharp.Tests/StatusFixture.cs [hunk 2]
```diff
@@ -103,20 +103,22 @@ public void RetrievingTheStatusOfADirectoryThrows()
             }
         }
 
-        [Fact]
-        public void CanRetrieveTheStatusOfTheWholeWorkingDirectory()
+        [Theory]
+        [InlineData(false, 0)]
+        [InlineData(true, 5)]
+        public void CanRetrieveTheStatusOfTheWholeWorkingDirectory(bool includeUnaltered, int untrackedCount)
```
Comment (Therzok) [inline LibGit2Sharp.Tests/StatusFixture.cs]: Testing both unaltered and altered for status dirtiness.

File: LibGit2Sharp.Tests/StatusFixture.cs [hunk 3]
```diff
@@ -592,5 +596,43 @@ public void RetrievingTheStatusHonorsAssumedUnchangedMarkedIndexEntries()
...
+
+        [Fact]
+        public void UnalteredFilesDontMarkIndexAsDirty()
+        {
+            var path = SandboxStandardTestRepo();
+
+            using (var repo = new Repository(path))
+            {
+                repo.Reset(ResetMode.Hard);
+                repo.RemoveUntrackedFiles();
```
Comment (Therzok) [inline LibGit2Sharp.Tests/StatusFixture.cs]: Cleaning stuff first. Even untracked files mark the repo as dirty.

File: LibGit2Sharp/RepositoryStatus.cs (hunk 1 above)
Comment (Therzok) [inline LibGit2Sharp/RepositoryStatus.cs]: On the other hand, I'll be keeping in mind to offer the option to IncludeUntracked files which would be `true` by default.

File: LibGit2Sharp.Tests/StatusFixture.cs (hunk 2 above)
Comment (nulltoken) [inline LibGit2Sharp.Tests/StatusFixture.cs]: Shouldn't it be `unalteredCount`?

Comment (Therzok) [inline LibGit2Sharp.Tests/StatusFixture.cs]: Yeah, I'm an idiot, fixing that now.

File: LibGit2Sharp/RepositoryStatus.cs (hunk 1 above)
Comment (nulltoken) [inline LibGit2Sharp/RepositoryStatus.cs]: I think the status already deals with untracked files by default.

However, would that be a typo (for `IncludeUnaltered`), I'm not sure about setting it to `true` by default. Most of the time, a commit will only change a very small portion of the tracked files. I *fear* that retrieving the whole content of the for each status call may be a little expensive.

Comment (Therzok) [inline LibGit2Sharp/RepositoryStatus.cs]: No, I meant, giving the user the option to filter out Untracked files and keeping backwards compatibility by default.

Unmodified on true by default = nono.

Comment (nulltoken) [inline LibGit2Sharp/RepositoryStatus.cs]: The way I see it, you actually fixed a bug (n the sense that we weren't matching git.git behavior), didn't you? If that's the case, 'm not sure we should provide an option for this. 

However, maybe should we isolate the change (and its code dependencies) in a separate commit?

Comment (Therzok) [inline LibGit2Sharp/RepositoryStatus.cs]: Ah, the option to take out Untracked files is something else.

You can query the git commandline in two ways for status changes:
git status --porcelain
git status --porcelain -uno (no untracked)

That's for later, yes.

Comment (nulltoken) [inline LibGit2Sharp/RepositoryStatus.cs]: >Ah, the option to take out Untracked files is something else.

Dammit! I read it backwards.

>That's for later, yes.

:+1:

**Comment by Therzok** [comment]
@nulltoken can you slap travis around with a large trout?

**Comment by nulltoken** [comment]
@Therzok Travis seems to have recovered. Still needed?

**Comment by Therzok** [comment]
As long as it's green, I'm okay with it. Ready to merge. :8ball:

File: LibGit2Sharp/RepositoryStatus.cs (hunk 1 above)
Comment (jamill) [inline LibGit2Sharp/RepositoryStatus.cs]: `FileStatus` is a `Flag` enum - do you mean to use the `!=` comparison, or should we use the `Enum.HasFlag` to check for these instead?

Comment (Therzok) [inline LibGit2Sharp/RepositoryStatus.cs]: These are exclusive statuses. Unaltered and Ignored can only appear by themselves.

Comment (Therzok) [inline LibGit2Sharp/RepositoryStatus.cs]: Also, x & 0 == 0 always. So we can't use anything else on Unaltered.

File: LibGit2Sharp/RepositoryStatus.cs [hunk 4]
```diff
@@ -164,14 +171,21 @@ private void AddStatusEntryForDelta(FileStatus gitStatus, GitDiffDelta deltaHead
 
             StatusEntry statusEntry = new StatusEntry(filePath, gitStatus, headToIndexRenameDetails, indexToWorkDirRenameDetails);
 
-            foreach (KeyValuePair<FileStatus, Action<RepositoryStatus, StatusEntry>> kvp in dispatcher)
+            if (gitStatus == FileStatus.Unaltered)
```
Comment (jamill) [inline LibGit2Sharp/RepositoryStatus.cs]: Could we handle the `FileStatus.Unaltered` value in the same way we handle the other values (in the dispatcher dictionary)?

Comment (jamill) [inline LibGit2Sharp/RepositoryStatus.cs]: Ah - probably not if `FileStatus.Unaltered = 0`.

**Comment by nulltoken** [comment]
@jamill Are you happy to see this in now?

**Comment by jamill** [comment]
Looks good, thanks!

**Comment by nulltoken** [comment]
@Therzok :+1:

**Pull request merged**